### PR TITLE
Add functions to SPI

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

**Comment by electrum** [comment]
@dain suggested prefixing `RankingWindowFunction` and `ValueWindowFunction` with `Abstract`.

**Comment by electrum** [comment]
Add Javadoc to new interfaces

**Comment by electrum** [comment]
Should we have sub-packages for separate function types? What things need to be shared? `@SqlType`, `@Description`, others?

**Comment by electrum** [comment]
@cberner might have had concerns previously about moving `ParametricType`

**Comment by electrum** [comment]
Decide if `BigArray` classes should be in SPI and maybe improve Javadoc to make it clear they are arrays (not lists)

File: presto-spi/src/main/java/com/facebook/presto/spi/function/IntermediateInputFunction.java [hunk 1]
```diff
@@ -11,15 +11,16 @@
...
 import java.lang.annotation.Target;
 
-@Retention(RetentionPolicy.RUNTIME)
-@Target(ElementType.METHOD)
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Retention(RUNTIME)
+@Target(METHOD)
 public @interface IntermediateInputFunction
```
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/IntermediateInputFunction.java]: I don't think we should put this one in the SPI. This is just a bad feature that we should remove entirely

**Comment by cberner** [comment]
I skimmed through the scalar and aggregation annotation changes, those seem fine, although I had one comment

**Comment by cberner** [comment]
You should talk to @martint about putting `ParametricType` in the SPI, at one point we had discussed unifying it with `Type`

**Comment by martint** [comment]
> Decide if BigArray classes should be in SPI and maybe improve Javadoc to make it clear they are arrays (not lists)

Ideally, the SPI should be only interfaces and exceptions. Implementations or convenience utilities should be made available to plugins via modules they can depend on.

**Comment by martint** [comment]
> You should talk to @martint about putting ParametricType in the SPI, at one point we had discussed unifying it with Type

I think it's fine to do for now, but there should be no expectation that these interfaces are yet stable. In fact, all the function/type-related abstractions are very likely to change in the future as we cleanup the type system and analyzer.

**Comment by electrum** [comment]
Remove `AbstractGroupedAccumulatorState` and inline in ML plugin

**Comment by electrum** [comment]
Create a module `presto-bigarray` with package `com.facebook.presto.bigarray`

**Comment by electrum** [comment]
This is ready for final review. I integrated the comments from @martint and @dain.

I'll be adding Javadoc for some of the moved interfaces.

File: presto-spi/src/main/java/com/facebook/presto/spi/function/IntermediateInputFunction.java (hunk 1 above)
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/IntermediateInputFunction.java]: The end goal of this is to remove `getServices` from `Plugin`, which can't happen until we have a replacement for `FunctionFactory` in the SPI. @martint says it's fine to add for now (it is used by a private plugin at Facebook). We can remove it once it's removed from the engine.

**Comment by cberner** [comment]
Someone else should approve the window function commits (`Convert WindowIndex into an interface` and `Move WindowFunction into SPI`), as I'm not very familiar with window functions. The rest looks good though

**Comment by dain** [comment]
Looks good

**Pull request closed without merging**